src/lib/coda_base/ledger.ml

```diff
@@ -2,416 +2,510 @@ open Core
+  [@@deriving sexp, bin_io]
+
+  let transaction : t -> Transaction.t Or_error.t =
+   fun {varying; _} ->
     let open Or_error.Let_syntax in
-    let%bind proposer_reward, emptys1, receiver_update =
-      match fee_transfer with
-      | None -> return (coinbase_amount, [], None)
-      | Some (receiver, fee) ->
-          (* This assertion will pass because of how coinbase is produced by Ledger_builder.apply_diff *)
-          assert (not @@ Public_key.Compressed.equal receiver proposer) ;
-          let fee = Amount.of_fee fee in
-          let%bind proposer_reward =
-            error_opt "Coinbase fee transfer too large"
-              (Amount.sub coinbase_amount fee)
-          in
-          let receiver_location, receiver_account, emptys =
-            get_or_initialize receiver
-          in
-          let%map balance = add_amount receiver_account.balance fee in
-          ( proposer_reward
-          , emptys
-          , Some (receiver_location, {receiver_account with balance}) )
-    in
-    let proposer_location, proposer_account, emptys2 =
-      get_or_initialize proposer
-    in
-    let%map balance = add_amount proposer_account.balance proposer_reward in
-    set t proposer_location {proposer_account with balance} ;
-    Option.iter receiver_update ~f:(fun (l, a) -> set t l a) ;
-    {Undo.coinbase= cb; previous_empty_accounts= emptys1 @ emptys2}
-
-  (* Don't have to be atomic here because these should never fail. In fact, none of
-   the undo functions should ever return an error. This should be fixed in the types. *)
-  let undo_coinbase t
-      { Undo.coinbase= {proposer; fee_transfer; amount= coinbase_amount}
-      ; previous_empty_accounts } =
-    let proposer_reward =
-      match fee_transfer with
-      | None -> coinbase_amount
-      | Some (receiver, fee) ->
-          let fee = Amount.of_fee fee in
-          let receiver_location =
-            Or_error.ok_exn (location_of_key' t "receiver" receiver)
-          in
-          let receiver_account =
-            Or_error.ok_exn (get' t "receiver" receiver_location)
-          in
-          set t receiver_location
-            { receiver_account with
-              balance=
-                Option.value_exn
-                  (Balance.sub_amount receiver_account.balance fee) } ;
-          Amount.sub coinbase_amount fee |> Option.value_exn
+    match varying with
+    | User_command tr ->
+        Option.value_map ~default:(Or_error.error_string "Bad signature")
+          (UC.check tr.common.user_command) ~f:(fun x ->
+            Ok (Transaction.User_command x) )
+    | Fee_transfer f -> Ok (Fee_transfer f.fee_transfer)
+    | Coinbase c -> Ok (Coinbase c.coinbase)
+end
+
+(* someday: It would probably be better if we didn't modify the receipt chain hash
+ in the case that the sender is equal to the receiver, but it complicates the SNARK, so
+ we don't for now. *)
+let apply_user_command_unchecked ledger
```

Comment:
I assume you didnt change the logic of any of these functions yes? just a bad diff?

Reply:
We just lifted these functions out of a functor